source-code/core2/src/messages/query.ts

```diff
@@ -0,0 +1,15 @@
+import type { Message } from "./schema.js"
+
+// TODO
+// 1. Query builder?
+// 2. Filtering etc how?
+//
+// inspiration: Prisma CRUD https://www.prisma.io/docs/concepts/components/prisma-client/crud
+export type MessagesQueryApi_1 = {
+	create: (message: Message) => void
+	get: () => Message[]
+	update: () => void
+	delete: () => void
+}
```

Comment:
## How to design Query API? 

### Problem 

We need a query API.

(Don't think about reactivity for now, that's a separate issue [#1122](https://github.com/inlang/inlang/issues/1122))

Reply:
Actually I love this prisma inspired API.

```ts
inlang.messages.query.create({ id: "myMessageId", languageTag: "en", pattern: [{"type": "Text", value:  "Hello World" }]})

const message = inlang.messages.query.get({ where: { id: "myMessageId", languageTag: "en" })

inlang.messages.query.update({ where: { id: "myMessageId", languageTag: "en" }, data: { pattern: [{"type": "Text", value:  "Hello World" }] } })

inlang.messages.query.delete({ where: { id: "myMessageId", languageTag: "en" } })
```

This brings us to the question of what is the return type of get?
To stay with prisma, they let you define specific properties as unique. So you could use `getUnique({ where: { id: "..." } })` to get only one return.  If you would use it in our case this wouldn't work because the combination of languageTag and id is unique. They also have `getFirst()` with returns also one and `getMany()` with returns the array. 

Personally, I do not like that there are this many variations. What do you think?

Do I get all when just using `inlang.messages.query.get()`?

Reply:
There need to be filter options in the crud functions. In my opinion it returns either `Message` or `Message[]` based on that you have chosen:

- only `id` provided -> `Message[]`
- `id` & `languageTag` -> `Message`
- `languageTag` & `data.pattern["type"]: "Text"` -> `Message[]`

There is no need to have this many functions as of now; controlling the returned results can/should be done by apps. 
We can potentially add them later if the apps have an extreme urge.

Do we want those query options and not just a regular filter? -> `{where: { /**/ }}, {include: {/**/}}, {select: {/**/}}, ...` syntax? https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#model-query-options

Reply:
> @felixhaeberle Do we want those query options and not just a regular filter? -> {where: { /**/ }}, {include: {/**/}}, {select: {/**/}}, ... syntax? https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#model-query-options

I don't understand the question. Please elaborate. 

> In my opinion it returns either Message or Message[] based on that you have chosen:

If you establish and enforce that each message with a `id-languageTag` combination can only exists once, we can distinguish if `get` returns a single message or an array. 

BUT if we introduce variants (A/B testing etc), the `get` function would be breaking because the unique combination would become `id-languageTag-variant`. 

Prisma gets around it by having `get` and `getMany`, `create`, `createMany`. Separating the query ops in single and many future proofs the API.

Reply:
👍 `get` and `getMany`, `create`, `createMany`. I do not love the variation of queries, but I think those set very distinct expectations and that is a good think. (I don't know if we need createMany, do you think of bulk translations?)

Reply:
> (I don't know if we need createMany, do you think of bulk translations?) 

For consistency, would be good to have `createMany` as well then.

Reply:
> I don't understand the question. Please elaborate.

@samuelstroschein Elaborated in person

Are we still agreeing with `get` and `getMany`, `create`, `createMany` – or does a more simpler API like `messages.list()` make sense?

Reply:
1. Is an `updateMany` also needed for consistency?
2. Don't we need an `upsert` function?
3. I see the use case where we want to push multiple changes (some are updates, some creates); therefore a `upsertMany` would be interesting.

Reply:
I think it makes more sense to make one PR and comment on that.

Reply:
@NilsJacobsen give a reason

Reply:
We gonna have multiple pretty similar starting points that are inspired by prisma. If you want I can hand in mine. Not sure if that is the most effective way of doing it.

Reply:
Okay, then you propose a concrete API in a PR? @felixhaeberle and @NiklasBuchfink you?

Reply:
I don't need to propose a full API. Due to different expected returns, I agree with `get` and `getMany`. 

I just wanted to bring to the discussion that `upsert` might still be helpful. If there should be an `updateMany` or `upsertMany` should be discussed if we see use cases for that.

For example, when applying multiple changes, do we need all three `updateMany`,  `createMany`, and `upsertMany`?